rts/lra/service-base/src/main/java/io/narayana/lra/Current.java

```diff
@@ -64,6 +72,67 @@ public static Object getState(String key) {
         return null;
     }
 
+    private static String getParents(URI uri) {
+        String query = uri.getQuery();
+
+        if (query != null) {
+            for (String nvpair : query.split(QUERY_PAIR_SEPARATOR)) {
+                if (nvpair.startsWith(PARENT_LRA_PARAM_NAME + QUERY_FIELD_SEPARATOR)) {
+                    return nvpair.split(QUERY_FIELD_SEPARATOR)[1];
+                }
+            }
+        }
+
+        return null;
+    }
+
+    // construct the LRA URI including the parent hierarchy as a query parameter
+    public static URI buildFullLRAUrl(String baseURI, URI parentId) throws URISyntaxException {
+        // is the parent part of a hierarchy
+        String parents = Current.getParents(parentId); // gets the hierarchy form the query param
+        // we have the hierarchy so remove the query parameter
+        String gParent = new URI(parentId.getScheme(),
+                parentId.getAuthority(),
+                parentId.getPath(),
+                null, // skip the query string
+                parentId.getFragment())
+                .toASCIIString();
+
+        if (parents != null) {
+            gParent += parents + ","; // , separated list of the hierarchy
+        }
+
+        return UriBuilder.fromUri(baseURI).queryParam(PARENT_LRA_PARAM_NAME, gParent).build();
+    }
+
+    // given a URL extract the immediate parent of
+    public static String getFirstParent(URI parent, boolean decode) throws UnsupportedEncodingException {
+        String query = parent == null ? null : parent.getQuery();
+
+        if (query != null) {
+            for (String param : query.split(QUERY_PAIR_SEPARATOR)) {
+                if (param.startsWith(PARENT_LRA_PARAM_NAME + QUERY_FIELD_SEPARATOR)) {
+                    String parents = param.split(QUERY_FIELD_SEPARATOR, 2)[1];
+
+                    // parents is a comma separated list of parents (the first one is the direct parent)
+                    if (parents != null) {
+                        String[] pa = parents.split(",");
+
+                        if (pa.length > 0) {
+                            if (decode) {
+                                return URLDecoder.decode(pa[0], "UTF-8");
+                            }
```

Comment:
is it correct to return null if `decode` is false? In that case, this can be refactored to not contain the parameter.